### DISCUSSION TITLE
Change the semanticCommitType in lockFileMaintenance based on dep types

### DISCUSSION THREAD

**Discussion by ryancausey** [renovatebot/renovate#19336] [Request Help] [answered]
I have a use case where I want to opt into lockfile maintenance, but I want to control the semantic commit type based on the underlying dep types being updated. For example:

1. if lockfile maintenance is being performed, and the only updates being made are to dev dependencies, I want the sematic commit type to be "chore".
2. if lockfile maintenance is being performed, and the updates include updates to the main dependencies and/or their required packages, I want the semantic commit type to be "fix".

I've messed around with a couple different package rules I thought might work, but it doesn't seem to be sticking. Here's the last iteration of the renovate.json file:

<details><summary>renovate.json</summary>
<p>

```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
        "config:base",
        ":semanticCommits"
    ],
    "rangeStrategy": "replace",
    "lockFileMaintenance": {
        "enabled": true,
        "schedule": [
            "at any time"
        ]
    },
    "packageRules": [
        {
            "matchUpdateTypes": [
                "lockFileMaintenance"
            ],
            "matchDepTypes": [
                "devDependencies",
                "dev-dependencies",
                "dev"
            ],
            "semanticCommitType": "chore"
        },
        {
            "matchUpdateTypes": [
                "lockFileMaintenance"
            ],
            "matchDepTypes": [
                "dependencies",
                "extras"
            ],
            "semanticCommitType": "fix"
        }
    ]
}
```

</p>
</details>

Here's a link to the repo I was testing with: https://gitlab.com/rccausey/test-lockfile-maintenance-package-rules

I had forced the dev dependency `black` to be behind on versions in the lock file as well as the main dependency `jsonpath-ng`. I was expecting this to see the main dependency needing an update and match on the last package rule. Is what I'm looking for possible?

**Comment by rarkins** [comment, marked as answer]
Renovate doesn't have any awareness of the deps being updated in lock file maintenance. It essentially just gets the package manager such as `npm` to completely recreate the lock file

**Reply by ryancausey** [reply]
Ah, I see. So the package rules trying to match on dep types will have no effect. And, if I understand correctly [from this discussion](https://github.com/renovatebot/renovate/discussions/19285#discussioncomment-4330717), lock file maintenance is the only way to update transitive dependencies without the direct dependency needing to change. Which I think means my other thought of using package rules to group all dev dependencies into one update won't really work to ensure transitive dependencies are also up to date. Is that correct?

**Reply by rarkins** [reply]
Transitive dependencies often get updated at the same time as direct dependencies.

Going back a step, do you have a technical reason for keeping ranges in your `package.json` instead of using pinned versions? e.g. are you writing a library?

**Reply by ryancausey** [reply]
In some cases, yes there is a library that is being maintained where ranges make sense (I think?). In other cases it's a deployed project and I can probably pin them instead. Does pinning make a difference as far as transitive dependency updates are concerned?

I'm mainly trying to ensure that the lock files are maintained and that I can have different rules for libraries vs deployed projects:

* If it's a library, lockfile maintenance is a purely development focused affair so I want to make the semantic commit type `chore` so semantic release bot doesn't cut a new release for it. This is straightforward and easy to configure by using the `matchUpdateTypes` package rule.
* If it's a deployed project, lockfile maintenance might update dependencies that are used in production if they are main dependencies, or transitive dependencies of main dependencies, instead of dev ones. In this case I was looking to be able to switch the semantic commit type to `fix` so that semantic release bot will actually release and deploy the changes. This doesn't seem possible when using `lockFileMaintenance`.

**Reply by rarkins** [reply]
For libraries, it usually makes sense to keep ranges. Otherwise, I recommend you pin where you can, because it not only simplifies updates but also gives better visibility to you and teammates. e.g. you can look at `package.json` and know exactly what you're using instead of needing to dig into lock files.

If you are keeping ranges, you could (for example), group together all non-major updates into a single PR. They will be applied one by one, and hopefully keep changes just to those in the list in the PR.

When you use lock file maintenance - which is also a good idea - it lets the package manager update *everything*, and it's up to you to look into the lock file to figure out what it is.

**Reply by ryancausey** [reply]
So my strategy is now:

* For all deployment projects that actually deploy something:
    * Set [rangeStrategy](https://docs.renovatebot.com/configuration-options/#rangestrategy) to `pin`.
    * Group all dev dependencies into one PR and make the semantic commit type `chore` using [packageRules](https://docs.renovatebot.com/configuration-options/#packagerules).
    * For all other dependencies keep the default [:semanticPrefixFixDepsChoreOthers](https://docs.renovatebot.com/presets-default/#semanticprefixfixdepschoreothers) behavior.
    * Do not enable [lockFileMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance).
* For library projects that are re-used code.
    * Set [rangeStrategy](https://docs.renovatebot.com/configuration-options/#rangestrategy) to `replace`.
    * Enable [lockFileMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance).
    * Make the semantic commit type `chore` for `lockFileMaintenance` using [packageRules](https://docs.renovatebot.com/configuration-options/#packagerules).
    * Keep the default [:semanticPrefixFixDepsChoreOthers](https://docs.renovatebot.com/presets-default/#semanticprefixfixdepschoreothers) behavior for other dependency updates.

Since it seems like in the pin situation, the transitive dependencies would be updated anytime the pinned dependency is updated, I think it should be OK to not have `lockFileMaintenance` enabled.
